### ISSUE TITLE
Add unused variable and parameter compiler flags

### ISSUE THREAD

**Issue by aprilnovak** [openmc-dev/openmc#2345]
The `-Wunused-parameter` and `-Wunused-variable` compiler flags warn if detecting an unused parameter (e.g. function argument) or an unused variable. I think these are very helpful in (i) identifying function arguments that can be simplified, (ii) detecting code errors/bad logic. For example, there's a few places in the code with statements like this:

```
if (const auto* ptr = dynamic_cast<const SurfaceXPlane*>(&surf2))
```

where `ptr` is unused. This if-statement I think will always evaluate to true, whereas the original author probably intended to actually check the result of the dynamic cast.

**Comment by paulromano** [comment]
> This if-statement I think will always evaluate to true

I don't agree with this — if the `dynamic_cast` fails, `ptr` will be assigned a null pointer that evaluates to false. Nevertheless, your point about `ptr` being unused stands.

**Comment by aprilnovak** [comment]
Ah, you're right - sorry about that, my mistake on the interpretation!
